File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/H._Время_выходить/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(V + E)
    Memory Complexity O(V + E)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    adj_list: dict[int, list[int]] = dict()
    for _ in range(m):
        source, destination = map(int, sys.stdin.readline().rstrip().split())
        if source not in adj_list: adj_list[source] = []
        adj_list[source].append(destination)
    for vertex, edges in adj_list.items():
        edges.sort(reverse=True)
    enter: list[int] = [0] * (n + 1)
    leave: list[int] = [0] * (n + 1)
    colors: list[str] = ['white'] * (n + 1)
    call_stack: list[int] = [1]
    current_time: int = -1
    while call_stack:
        vertex = call_stack.pop()
        if colors[vertex] != 'black': current_time += 1
        if colors[vertex] == 'white':
            enter[vertex] = current_time
            colors[vertex] = 'grey'
            call_stack.append(vertex)
            for neighbor in adj_list.get(vertex, []):
                if colors[neighbor] == 'white':
                    call_stack.append(neighbor)
        elif colors[vertex] == 'grey':
            colors[vertex] = 'black'
            leave[vertex] = current_time
    for vertex in range(1, n + 1):
        sys.stdout.write(f'{enter[vertex]} {leave[vertex]}\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/H._Время_выходить/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>
#include <algorithm>


void solution() {
    /*
    Time Complexity O(V + E)
    Memory Complexity O(V + E)
    */
    int n, m;
    std::cin >> n >> m;
    std::unordered_map<int, std::vector<int>> adjList;
    for (int i = 0; i < m; ++i) {
        int source, destination;
        std::cin >> source >> destination;
        adjList[source].push_back(destination);
    }
    for (auto& pair : adjList) {
        std::sort(pair.second.begin(), pair.second.end(), std::greater<int>());
    }
    std::vector<int> enter (n + 1, 0);
    std::vector<int> leave (n + 1, 0);
    std::vector<std::string> colors (n + 1, "white");
    std::vector<int> callStack = {1};
    int curTime = -1;
    while (!callStack.empty()) {
        int& vertex = callStack[callStack.size() - 1];
        callStack.pop_back();
        if (colors[vertex] != "black") ++curTime;
        if (colors[vertex] == "white") {
            colors[vertex] = "grey";
            enter[vertex] = curTime;
            callStack.push_back(vertex);
            for (int& neighbor : adjList[vertex]) {
                if (colors[neighbor] == "white") {
                    callStack.push_back(neighbor);
                }
            }
        } else if (colors[vertex] == "grey") {
            colors[vertex] = "black";
            leave[vertex] = curTime;
        }
    }
    for (int vertex = 1; vertex <= n; ++vertex) {
        std::cout << enter[vertex] << " " << leave[vertex] << "\n";
    }
}


int main() {
    solution();
}